### ISSUE TITLE
JNA quick example

### ISSUE THREAD

**Issue by Paul776** [Grover-c13/PokeGOAPI-Java#493]
I started writing the fix for that signature stuff. It requires java (JNA easier than JNI) to read some bloody c code. If you guys want to get started here is how

1) Grap the encrypt.c and .h (somewhere on the subreddit)
2) make
3) grap the .so (or dll if you are on windows)

``````
public class Test {

  public interface CTest extends Library {
    public int encrypt(byte[] input, long size, byte[] iv, long size2, byte[] output, int[] output size)
  }
  public static void main(String ... args) {
    CTest ctest = (CTest) Native.loadLibrary("encrypt", CTest.class);
    System.out.println(ctest.encrypt(new byte[3], 3, iv, 32, null, new int[50000]));
    //System.out.println(new Encrypt().encrypt(new byte[3], 3, iv, 32, null, 32));
  }  
}
```;

It should output 0

the real input are
(raw_signature_proto_buf, size, iv, 32, output, output size pointer)

I'm off to bed hope I see a PR tomorrow morning 👍 
``````

**Comment by DanippDm5vzDS8HWnc9fap** [comment]
I am working on re-scripting the encryption to java to a jar. Probably you could skip the jnit part at this moment and let me see if the java implementation of the encryption works.

**Comment by illusion12** [comment]
Ideally you would rewrite the encryption code in java. Using their library isn't very optimal. JNA/JNI is messy IMO

PS: do you happen to have encrypt.c and encrypt.h?

**Comment by DanippDm5vzDS8HWnc9fap** [comment]
@illusion12  working on it but probably cant share code here maybe only the resulting jar.

**Comment by illusion12** [comment]
Why not?

**Comment by DanippDm5vzDS8HWnc9fap** [comment]
just like the pokemondev they dont directly share the c code any more due to the legal issue. It is propitiatory.

**Comment by LoungeKatt** [comment]
@DanippDm5vzDS8HWnc9fap They established a site with the entire purpose of sharing the c code. They just don't directly host it without first having to read a bunch of disclaimers.

**Comment by DanippDm5vzDS8HWnc9fap** [comment]
@LoungeKatt  where is the site sharing c code? couldnt find it any more. do you have the link?

**Comment by avelizmu** [comment]
Here you go
http://pgoapi.com/

**Comment by LoungeKatt** [comment]
Now is it cool if we use slack for the conversation and this goes back to generating code to implement?

**Comment by DanippDm5vzDS8HWnc9fap** [comment]
almost done with the rewriting. they also provide test data. good to have.

**Comment by Paul776** [comment]
I have a working JNA code (and actually the all flow working) but it is mess indeed
hmm good luck to rewrite the encrypt.c into java! it's impossible so much code :/

**Comment by DrBrad** [comment]
Paul776 If you need any help I'm good with Java and would be willing to help out.

**Comment by DanippDm5vzDS8HWnc9fap** [comment]
@Paul776  semi-automated. almost done. testing.

**Comment by DrBrad** [comment]
Is this all that needs to be done for the api to actually work again?

**Comment by DanippDm5vzDS8HWnc9fap** [comment]
@DrBrad nope. compile proto. generate signature for every message envelope. encripted. put it into unknown6.

**Comment by ramarro123** [comment]
i don't think will be added to this api anytime soon unfortunately :(

**Comment by illusion12** [comment]
@Paul776 it is not impossible. Plus using JNA is slow. I have been able to implement most of the Signature proto (minus the final hash using the encryption) but I am too tired to finish it tonight. I will probably have this all implemented tomorrow at some point.

Assuming this xxhash library I found functions correctly. I may have to do some testing on it. If it fails, I will write my own version.

**Comment by DrBrad** [comment]
@ramarro123 Do you know of any other Java PokemonGo-API's out currently? As you could use another api to help with this one.

**Comment by jsl303** [comment]
@illusion12 Just curious. Is it possible to import C library using jni? I read jni is faster than jna?

**Comment by Paul776** [comment]
Of course it is possible. But it's true it is not the relate solution.

 1) it's complicated
 2) slow
 3) Might not work well for the Android people

Better try to translate with a tool the .c to .java and use it natively

@DanippDm5vzDS8HWnc9fap Any luck in the end?

I have a working JNA implementation. I can call the C code. I get things back but not exactly he right things.
I think it's because my inputs are wrong. The python script as a d2h which convert the double location to a kind of hex but I haven't been able to do this correctly in java
it gives me back xxHashes negative :/

**Comment by DanippDm5vzDS8HWnc9fap** [comment]
@Paul776  completed. tested with varying random input of size 200, 256, 500 and seems to be correct. but i dont have dump to fully verify

**Comment by clangdev** [comment]
Actually I fail at the make-step (ld -lencrpyt missing) and using a windows dll I found in chat I get -1 as result.

**Comment by jabbink** [comment]
For JNA, see https://github.com/Grover-c13/PokeGOAPI-Java/pull/494

feel free to make that thing actually work as right now either the encryption thing or the xxhash'ing is not working correctly.

**Comment by Paul776** [comment]
@DanippDm5vzDS8HWnc9fap 
http://imgur.com/a/338XO

It is working perfectly (meaning I have a valid final output)
but the making the request I get 0 pokemons back

I will gave up with JNA it's not the right solution. Could you share your converted java code? I can give it a try right now with a valid request

**Comment by jabbink** [comment]
https://github.com/Grover-c13/PokeGOAPI-Java/pull/494

Has Java code

**Comment by Paul776** [comment]
@jabbink 
I have hooked my JNA imp with your code, and it return 0 pokemons as well
might be something wrong in the way we build  the requests

**Comment by vmarchaud** [comment]
merged #494 so i close this issue ;)
